### PR TITLE
Feature/artv3/deprecateforall(begin,end)

### PR DESCRIPTION
**Pull request by artv3**
@ajkunen , @rhornung67,  I added deprecation macros to the forall methods of the form forall(begin,end). Should I do the same for forallN? I currently have one but it spews out many warnings as it is being used in unit tests.

### CODE REVIEW

**Comment by rhornung67** [comment]
@artv3 did you check the compiler output for this with the clang compiler? I've noticed that it can emit a lot of warnings related to the deprecation macro. That's OK as long as it only does this for actual use of deprecated features. Also, if we have a lot of warnings coming out of the tests and/or examples, these should be converted to use segments explicitly.

**Comment by davidbeckingsale** [comment]
I agree with @rhornung67 regarding updating tests to avoid the warnings. The stuff coming from clang though looks like an error - might be to do with the placement of the deprecation macro.

**Comment by willkill07** [comment]
If we deprecate this, should we move span and make_span out of internal to provide a stop-gap for those who were passing in raw pointers with size offsets?

**Comment by willkill07** [comment]
Also, any attributes should follow the parameter list

http://en.cppreference.com/w/cpp/language/function

e.g.
```
void forall(T begin, T end, Body b) RAJA_DEPRECATE(“Message here”) {

}
```

**Comment by artv3** [comment]
I fixed the location of the marcos (it now works with clang). @willkill07 , I tried moving the macros to where you suggested but I ended up getting build errors. Also, I have paired a macro with forallN and since there are unit test it is currently spewing out warnings when building.

**Comment by davidbeckingsale** [comment]
Hey @artv3 - can you add a CMake option `ENABLE_DEPRECATED_TESTS` which is Off by default and use that to guard if tests with deprecated features get compiled.

File: test/unit/cuda/test-reduce-loc.cpp [hunk 1]
```diff
@@ -167,7 +167,7 @@ CUDA_TYPED_TEST_P(ReduceCUDA, generic)
       reduce::detail::ValueLoc<double> randval(droll, index);
       applier::updatedvalue(dvalue, randval, dcurrent);
 
-      forall<cuda_exec<block_size>>(0, TEST_VEC_LEN, [=] __device__(int i) {
+      forall<cuda_exec<block_size>>(RAJA::RangeSegment(0, TEST_VEC_LEN), [=] __device__(int i) {
```
Comment (davidbeckingsale) [inline test/unit/cuda/test-reduce-loc.cpp]: Is the `RangeSegment` alias going to be deprecated for `TypedRangeSegment` any time soon?

Comment (artv3) [reply test/unit/cuda/test-reduce-loc.cpp]: I don't see it on the TODO list. @rhornung67, will we be keeping the RangeSegments?

Comment (rhornung67) [reply test/unit/cuda/test-reduce-loc.cpp]: I think we should leave the type aliases for RangeSegment and RangeStrideSegment in place for a while at least. We should discuss as a group whether to deprecate and, if so, when.

Comment (davidbeckingsale) [reply test/unit/cuda/test-reduce-loc.cpp]: Sounds good to me.

**Comment by artv3** [comment]
This branch is good to go :-)

**Comment by davidbeckingsale** [comment]
Have you tried building with deprecated tests enabled. From the diff it doesn't look like the variable will ever be set. You should add a `RAJA_ENABLE_DEPRECATED_TESTS` variable to `config.hpp.in`, setup the same way as the others.

**Comment by artv3** [comment]
@davidbeckingsale, that was simpler, thanks! This should work now. Before merging, we should squash my PRs. I think I may be committing too much.

**Comment by davidbeckingsale** [comment]
My only question is whether we want to deprecate forallN right now, or wait a little longer. @rhornung67 ?

**Comment by rhornung67** [comment]
@davidbeckingsale or @artv3 will one of you ping the folks that are using forallN and ask  them if they're OK with deprecating it in the upcoming release?  Based on their responses, we'll decide. Also, please make sure that there are no spurious warning messages generated by compilers if it's not used. I want to avoid the issues we had in a recent release.  If you have questions, please let me know.  Thanks.

File: include/RAJA/pattern/forallN.hpp [hunk 2]
```diff
@@ -116,7 +116,9 @@ struct ForallN_Executor<0> {
  */
 
 template <typename POLICY, typename BODY, typename... ARGS>
-RAJA_INLINE void forallN_policy(ForallN_Execute_Tag,
+RAJA_DEPRECATE("ForallN will be deprecated in next release")
+RAJA_INLINE
+void forallN_policy(ForallN_Execute_Tag,
```
Comment (davidbeckingsale) [inline include/RAJA/pattern/forallN.hpp]: Why don't you deprecate the "raw" forallN call?

File: test/unit/cpu/test-reductions.cpp [hunk 3]
```diff
@@ -413,6 +413,7 @@ class NestedReductionCorrectnessTest : public ::testing::Test
 };
 TYPED_TEST_CASE_P(NestedReductionCorrectnessTest);
 
+#ifdef RAJA_DEPRECATED_TESTS
```
Comment (davidbeckingsale) [inline test/unit/cpu/test-reductions.cpp]: Use `#if defined()`

**Comment by artv3** [comment]
@davidbeckingsale , shall I merge?

**Comment by davidbeckingsale** [comment]
Can you update the travis so that at least one compiler builds the deprecated tests?

**Comment by artv3** [comment]
@davidbeckingsale , @rhornung67 , I updated Travis to test GCC49 with deprecated tests. Can I get another PR approval?

**Pull request merged**